### PR TITLE
Initial AIS implementation

### PR DESCRIPTION
**Pull request by gskjold**
### What is the linked issue for this pull request:
[comment]: # (Add the ticket number behind QMS-# )

QMS-#586

### What you have done:
[comment]: # (Describe roughly.)
Implemented a new realtime source that parses AIS data and shows them on the map. Mostly a copy/paste from GPS tether and OpenSky sources.

### Steps to perform a simple smoke test:
[comment]: # (List the steps.)

1. Add new AIS source
2. Enter hostname and port, eks 153.44.253.27 port 5631 (Public source https://www.kystverket.no/en/navigation-and-monitoring/ais/access-to-ais-data/)
3. Click connect
4. Watch how ships appears on map, mostly around western part of Norway.

### Does the code comply to the coding rules and naming conventions [Coding Guidelines](https://github.com/Maproom/qmapshack/wiki/DeveloperCodingGuideline):

- [x] yes

### Is every user facing string in a tr() macro?

- [x] yes

### Did you add the ticket number and title into the changelog? Keep the numeric order in each release block.

- [x] yes, I didn't forget to change changelog.txt

### CODE REVIEW

**Comment by kiozen** [comment]
Nice! I will do a review the next days.

File: src/qmapshack/realtime/ais/CRtAisInfo.cpp [hunk 1]
```diff
@@ -0,0 +1,591 @@
...
+uint32_t CRtAisInfo::getInt(const QByteArray& data, int start, int count)
+{
+    int mask = 0x3f;
+
+    int end = start + count;
+
+    int from = start / 6;
+    int to = end / 6;
+
+    int from_mask_ignore_bits = start % 6;
```
Comment (kiozen) [inline src/qmapshack/realtime/ais/CRtAisInfo.cpp:498]: I guess this code is copy-n-paste from another project. Do we have the copy right for that? Do we need to mention the original author?

Well and... nameing conventions.  We do camelCase for variables.

File: src/qmapshack/realtime/ais/CRtAisInfo.cpp [hunk 2]
```diff
@@ -0,0 +1,591 @@
+/**********************************************************************************************
+    Copyright (C) 2018 Oliver Eichler <oliver.eichler@gmx.de>
```
Comment (kiozen) [inline src/qmapshack/realtime/ais/CRtAisInfo.cpp:2]: Don't be shy. Don't hide yourself. You had the work, you deserve the copyright.

Applies to all the other files, too.

File: src/qmapshack/realtime/ais/CRtAis.cpp [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    return ships.count();
+}
+
+bool CRtAis::getShowNames() const
+{
+    QMutexLocker lock(&IRtSource::mutex);
+    return showNames;
+}
+
+CRtAis::ship_t* CRtAis::getShipByKey(const QString& key, bool& ok)
```
Comment (kiozen) [inline src/qmapshack/realtime/ais/CRtAis.cpp:103]: This pointer gives me a shiver. Yes, it's never a nullptr. But returning a `const ship_t&` would express it much better and wouldn't make me scream for a pointer check every time you use it ;)

**Review by kiozen — CHANGES_REQUESTED** [review]
Nicely done. Just a few minor things.

File: src/qmapshack/realtime/ais/CRtAisInfo.cpp (hunk 1 above)
Comment (gskjold) [reply src/qmapshack/realtime/ais/CRtAisInfo.cpp:498]: I am the original author, but yes, copy paste from a java project I made probably 10+ years ago that decodes AIS :)

File: src/qmapshack/realtime/ais/CRtAis.cpp (hunk 3 above)
Comment (gskjold) [reply src/qmapshack/realtime/ais/CRtAis.cpp:103]: Hehe, got it, I will sort this one out :)

File: src/qmapshack/realtime/ais/CRtAisInfo.cpp (hunk 2 above)
Comment (gskjold) [reply src/qmapshack/realtime/ais/CRtAisInfo.cpp:2]: Came with the copy/paste from opensky and gpstether files, didn't really pay much attention to it :)

**Review by gskjold — COMMENTED** [review]
Committed some changes earlier that should cover it, I hope :)

**Review by kiozen — APPROVED** [review]

**Comment by kiozen** [comment]
Ready for merge? Or do you want to add stuff?

**Comment by gskjold** [comment]
I was thinking I could add dimensions and ship type to the pop-up, which both could be interesting. I'll see what I can do this weekend

**Comment by gskjold** [comment]
Added vessel type and dimension to the info box. Not sure if I nailed the lookup maps for the types, my qt/c++ is a bit rusty.. This will be my final addition to this feature for now, let me know if I should make any changes

**Comment by kiozen** [comment]
A bit nicer way to initialize the maps would look like that:
```    
QMap<quint8, QString> shipTypeMap = {
        {0, ""},
        {20, tr("Wing in ground")},
        {21, tr("Wing in ground") + ", " + tr("Hazardous category A")}
    };
```
The way you are doing it currently involves a lot of code that has to be processed. As it's a one time task it's not that crucial. But if you want to make it nice and shiny go ahead :smile:

**Comment by wthaem** [comment]
@gskjold: Please check the code for a crash:

* Open the realtime window
* Add OpenSky and AIS sources without activating them
* Remove OpenSky source 
* Remove AIS source

On Windows I get a crash at the last step.  Can be reproduced!

**Comment by kiozen** [comment]
@wthaem I tried it ob Linux and can't reproduce. Can you get a backtrace?

**Comment by wthaem** [comment]
Some additional info for described crash:

* Doesn't happen if not yet connected to server - thus, connect to server first and see vessels before removing source
* Backtrace:

~~~
Ausnahme ausgelöst bei 0x00007FFCD31D051E (Qt5Widgets.dll) in qmapshack_reldbg.exe: 0xC0000005: Zugriffsverletzung beim Lesen an Position 0xFFFFFFFFFFFFFFFF.

 	Qt5Widgets.dll!00007ffcd31d051e()	Unbekannt
>	qmapshack_reldbg.exe!CRtAisInfo::slotDisconnected() Zeile 145	C++
 	[Externer Code]	
 	qmapshack_reldbg.exe!main(int argc, char * * argv) Zeile 78	C++
 	[Externer Code]	
~~~

* Source code:

~~~
void CRtAisInfo::slotDisconnected()
{
    lineHost->setEnabled(true);
    spinPort->setEnabled(true);
    toolConnect->setChecked(false);
    toolConnect->setIcon(QIcon("://icons/32x32/Disconnected.png"));  <=========== line 145

    slotUpdate();

    autoConnect(5000);
}

~~~

**Comment by kiozen** [comment]
@gskjold This should be quite easy to fix:
```
CRtAisInfo::~CRtAisInfo()
{
    socket->disconnect();
}
```
This will disconnect all signals from the socket. By that the signals do not backlash while destructing all objects with the danger to trigger slots of already destroyed objects.

**Comment by gskjold** [comment]
> The way you are doing it currently involves a lot of code that has to be processed. As it's a one time task it's not that crucial. But if you want to make it nice and shiny go ahead

Sweet, looks much better :)


> @gskjold This should be quite easy to fix

Added!

**Comment by kiozen** [comment]
Nice! I guess we merge that stuff now. Thanks for the work!

**Comment by anesthesia79** [comment]
Hello,
i just want to mention that by compiling (fresh) i get following issue

[ 16%] Generating ui_IRoutinoDatabaseBuilder.h
/home/emi/Builds/QMapShack/src/qmapshack/tool/IRoutinoDatabaseBuilder.ui: Warning: Z-order assignment: 'verticalLayoutWidget' is not a valid widget.
[ 16%] G**enerating ui_IRtAisInfo.h**
/home/emi/Builds/QMapShack/src/qmapshack/realtime/ais/IRtAisInfo.ui: Warning: The name 'gridLayout' (QGridLayout) is already in use, defaulting to 'gridLayout1'.
/home/emi/Builds/QMapShack/src/qmapshack/realtime/ais/IRtAisInfo.ui: Warning: The name 'label' (QLabel) is already in use, defaulting to 'label1'.
/home/emi/Builds/QMapShack/src/qmapshack/realtime/ais/IRtAisInfo.ui: Warning: The name 'horizontalLayout' (QHBoxLayout) is already in use, defaulting to 'horizontalLayout1'.
/home/emi/Builds/QMapShack/src/qmapshack/realtime/ais/IRtAisInfo.ui: Warning: The name 'label_2' (QLabel) is already in use, defaulting to 'label_21'.
[ 16%] Generating ui_IRtGpsTetherInfo.h

QMapShack starts as expected.

OS: Linux Mint 21.1


HTH,

**Comment by wthaem** [comment]
Similar warnings when compiling with MSVC.

**Comment by kiozen** [comment]
fixed it

**Pull request merged**